### DISCUSSION TITLE
Using Pages inside Python Classes

### DISCUSSION THREAD

**Discussion by Slanderkin** [zauberzeug/nicegui#540] [Q&A] [answered]
How do you go about creating a page function inside of a python class?

When I attempt to make a page like this:
```python
class HomePage:

    def __init__(self):
        #some content here

    @ui.page('/')
    def index_page(self) -> None:
        #some content here
```

I get an error of value_error.missing - which looks like it gets confused by the self keyword being added

![image](https://user-images.githubusercontent.com/16721903/225773564-b4e52757-7f95-4584-9b7b-8ee4433ecadd.png)

Is it possible to include a page function in a class like this or should it be done some other way?

**Comment by rodja** [comment, marked as answer]
Interesting. I've never thought about that. But it's a really good question. FastAPI seems also unable to do this:

```py
class Test:

    @app.get('/test')
    def test(self):
        return 'does this work?'
```

Returns a json stating "value self is missing" instead of "does this work?". But it does not mean that NiceGUI should fail here too. Are there any good reasons why this would be a bad idea?

**Reply by falkoschindler** [reply]
I'm not sure about the expected behavior. Isn't the decorator executed during class definition? So the route is created independent of a `Test` object, which allows calling `test()` without instantiating `Test()`. Then what should `self` be?

I see two options:

1. Remove `self` (and optionally add a `@staticmethod` decorator). This makes `test()` a single (static) route that is created with the class. This, however, removes access to `self`.
2. Move the whole route definition into `__init__()`. That's what we usually do. This creates a new routes only when a `Test` object is instantiated. It also allows to access `self` from the surrounding `__init__` context.

**Reply by Slanderkin** [reply]
Are you saying that the code would look like this:
```python

class HomePage:
    def __init__(self):
        self.route = "/"

        @ui.page(self.route)
        def index_page(self) -> None:
            #some content here
```

My main goal with adding a page definition under a class was so that I could easily centralize all information in one place. I want to include an AG Grid on a page and it would easy to have a parent page class with **self.row_data** that can be used within a function defining a page with an AG Grid.

I see now that since the page decorator is executed at class definition that maybe it would be better to just edit the AG Grid's data at run time. So maybe have the grid itself defined as a member for the class and then update that and have that update reflected in the page.

I still think there are valid cases to have a page defined as a class function but I can probably work around it.
